bndtools.builder/src/org/bndtools/builder/classpath/BndContainerInitializer.java

```diff
@@ -488,7 +488,42 @@ private void addProjectEntry(IPath path, List<IAccessRule> accessRules, List<ICl
 
 		private IPath calculateSourceAttachmentPath(IPath path, File file) {
 			JarInfo info = getJarInfo(file);
-			return info.hasSource ? path : null;
+			IPath sourcePath = info.hasSource ? path : null;
+
+			if (sourcePath == null) {
+				sourcePath = trySourcesJar(file);
+			}
+
+			return sourcePath;
+		}
+
+		/**
+		 * try to see if there is file bsn-sources.jar e.g. as downloaded by the
+		 * Maven repos
+		 *
+		 * @param file
+		 * @return the bsn-sources.jar file if found, otherwise
+		 *         <code>null</code>
+		 */
+		private IPath trySourcesJar(File file) {
+
+			if (file.getName()
+				.endsWith(".jar")) {
+
+				try {
+					File withSources = new File(file.getParentFile(), file.getName()
+						.substring(0, file.getName()
+							.indexOf(".jar"))
```

Comment:
Should probably be a `lastIndexOf`. How about using a method like Apache Commons IO `FilenameUtils.getExtension()`?

Reply:
Thanks. I'll try to improve it. FilenameUtils is currently not a dependency at the moment and I don't want to introduce one just for this.